In [1]:
import pandas as pd 
import numpy as np 
from sklearn import tree
from sklearn import preprocessing


C:\Users\hp\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\hp\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\hp\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\hp\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


In [3]:
df=pd.read_csv("train.csv")
df.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [6]:
df.duplicated().sum()



0

In [55]:
le=preprocessing.LabelEncoder()
df['Sex']=le.fit_transform(df['Sex'])
df.head(2)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   

      Ticket     Fare Cabin Embarked  
0  A/5 21171   7.2500   NaN        S  
1   PC 17599  71.2833   C85        C

**TRAINING DECISION TREE MODEL:**

In [58]:
features=["Sex","Age","Fare"]

In [59]:
tree_model=tree.DecisionTreeClassifier(max_depth=8)
tree_model.fit(X=df[features],y=df["Survived"])


DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=8, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [60]:
with open("dtree.dot",'w') as f:
    f=tree.export_graphviz(tree_model,feature_names=["Sex","Age","Fare"],out_file=f);

Model Accuracy:

In [64]:
print(tree_model.score(X=df[features],y=df["Survived"])*100,"percentage")


87.06411698537683 percentage


**DECISION TREE:**

In [65]:
from sklearn.tree import export_text
r=export_text(tree_model,feature_names=features)
print(r)



|--- Sex <= 0.50
|   |--- Fare <= 48.20
|   |   |--- Fare <= 27.82
|   |   |   |--- Fare <= 25.70
|   |   |   |   |--- Fare <= 24.07
|   |   |   |   |   |--- Fare <= 10.48
|   |   |   |   |   |   |--- Fare <= 8.04
|   |   |   |   |   |   |   |--- Fare <= 6.99
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Fare >  6.99
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Fare >  8.04
|   |   |   |   |   |   |   |--- Age <= 19.00
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Age >  19.00
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- Fare >  10.48
|   |   |   |   |   |   |--- Fare <= 14.13
|   |   |   |   |   |   |   |--- Age <= 53.50
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Age >  53.50
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- Fare >  14.13
|   |   |   |   |   |   |   |--- Fare <= 15.37
|   |   |   |   |   

Inference:

Few rules:<br>
.Females with Fare>48.2 and age<8 not survived 


**PREDICTION:**

In [66]:
test=pd.read_csv("test.csv")
test['Sex']=le.fit_transform(test['Sex'])
df.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   

      Ticket     Fare Cabin Embarked  
0  A/5 21171   7.2500   NaN        S  
1   PC 17599  71.2833   C85        C

In [68]:
test_features=["sex","Age","Fare"]
test_pred=tree_model.predict(X=test[features])


In [70]:
predicted_output=pd.DataFrame({"passengerid":test["PassengerId"],"Survived":test_pred})
predicted_output.to_csv("output.csv",index=False)